src/ast/passes/macro_expansion.cpp

```diff
@@ -166,127 +225,179 @@ void MacroExpander::visit(Map &map)
+          err << "Call to " << name << "() "
+              << "has arguments that do not match any definition.";
+
+          // We can't say exactly what they intended here, so we break this into
+          // a suitable hint, pointing to the closest macro we could find.
+          if (macro->vargs.size() != args.size()) {
+            err.addContext(macro->loc)
+                << "The closest definition of " << macro->name
+                << "() has a different number of arguments. "
+                << "Expected: " << macro->vargs.size() << " but got "
+                << args.size();
+            return;
+          }
+
+          // Break down the mismatched arguments.
+          for (size_t i = 0; i < macro->vargs.size() && i < args.size(); i++) {
+            if (auto *mmap = macro->vargs.at(i).as<Map>()) {
+              if (args.at(i).is<Variable>()) {
+                auto &arg_err = args[i].node().addError();
+                arg_err << "Mismatched arg. " << macro->name
+                        << "() expects a map for arg " << mmap->ident
+                        << " but got a variable.";
+                arg_err.addContext(mmap->loc)
+                    << "This is the argument in the closest definition.";
+              } else {
+                auto &arg_err = args[i].node().addError();
+                arg_err << "Mismatched arg. " << macro->name
+                        << "() expects a map for arg " << mmap->ident
+                        << " but got an expression.";
+                arg_err.addContext(mmap->loc)
+                    << "This is the argument in the closest definition.";
+              }
+            } else if (auto *mvar = macro->vargs.at(i).as<Variable>()) {
+              if (args.at(i).is<Map>()) {
+                auto &arg_err = args[i].node().addError();
+                arg_err << "Mismatched arg. " << macro->name
+                        << "() expects a variable for arg " << mvar->ident
+                        << " but got a map.";
+                arg_err.addContext(mvar->loc)
+                    << "This is the argument in the closest definition.";
+              } else {
+                auto &arg_err = args[i].node().addError();
+                arg_err.addContext(args[i].node().loc)
+                    << "Mismatched arg. " << macro->name
+                    << "() expects a variable for arg " << mvar->ident
+                    << " but got an expression.";
+                arg_err.addContext(mvar->loc)
+                    << "This is the argument in the closest definition.";
+              }
+            }
+          }
+        });
+    if (!done) {
+      // This should not happen; add the error.
+      LOG(BUG) << done.takeError();
     }
+    return;
   }
 
-  if (auto it = macros_.find(name); it != macros_.end()) {
-    Macro *macro = it->second;
-
-    if (is_recursive_call(name, expr.node())) {
+  const auto *macro = *result;
+  for (size_t i = 0; i < stack_.size(); ++i) {
```

Comment:
As you mentioned, maybe we should allow this and just limit the number of recursive expansions but we can do that in a follow-up

Reply:
Yeah, I am just doing the overloading here. No intent to change the recursive behavior at this time.